File: backend/src/BagChronos.Domain/Services/VacationWorkflow.cs
using BagChronos.Domain.Entities;
using BagChronos.Domain.Enums;

namespace BagChronos.Domain.Services;

/// <summary>
/// State machine for vacation (and other Epic-4 workflowed) requests.
/// Allowed transitions:
///   Draft               → Submitted | Cancelled
///   Submitted           → PendingSubstitute | PendingManager | Cancelled
///   PendingSubstitute   → PendingManager | ReturnedForRevision | Cancelled
///   PendingManager      → PendingHr | Approved | Rejected | ReturnedForRevision
///   PendingHr           → Approved | Rejected
///   ReturnedForRevision → Submitted (resubmit) | Cancelled
///   Approved | Rejected | Cancelled → terminal (Approved/Cancelled may transition to Cancelled if before start)
/// </summary>
public static class VacationWorkflow
{
    public static void Submit(Request request, Guid actorId, DateTimeOffset at)
    {
        EnsureState(request, WorkflowState.Draft, WorkflowState.ReturnedForRevision);
        request.WorkflowState = request.SubstituteId is not null
            ? WorkflowState.PendingSubstitute
            : WorkflowState.PendingManager;
        request.Status = RequestStatus.Submitted;
        SyncDerived(request, at);
    }

    public static void AcceptSubstitute(Request request, Guid actorId, DateTimeOffset at)
    {
        EnsureState(request, WorkflowState.PendingSubstitute);
        EnsureActor(request.SubstituteId, actorId, "substitute");
        request.SubstituteAcceptedAt = at;
        request.WorkflowState = WorkflowState.PendingManager;
        SyncDerived(request, at);
    }

    public static void DeclineSubstitute(Request request, Guid actorId, DateTimeOffset at, string note)
    {
        EnsureState(request, WorkflowState.PendingSubstitute);
        EnsureActor(request.SubstituteId, actorId, "substitute");
        if (string.IsNullOrWhiteSpace(note))
        {
            throw new InvalidOperationException("A note is required when a substitute declines.");
        }
        request.WorkflowState = WorkflowState.ReturnedForRevision;
        request.Status = RequestStatus.Submitted;
        request.DecisionNote = note;
        SyncDerived(request, at);
    }

    public static void ManagerApprove(Request request, Guid managerId, DateTimeOffset at, string? note = null, bool requiresHrConfirmation = false)
    {
        EnsureState(request, WorkflowState.PendingManager);
        request.ApproverId = managerId;
        request.DecisionNote = note;
        request.DecidedAt = at;
        if (requiresHrConfirmation)
        {
            request.WorkflowState = WorkflowState.PendingHr;
            request.Status = RequestStatus.Submitted;
        }
        else
        {
            request.WorkflowState = WorkflowState.Approved;
            request.Status = RequestStatus.Approved;
        }
        SyncDerived(request, at);
    }

    public static void ManagerReject(Request request, Guid managerId, DateTimeOffset at, string? note = null)
    {
        EnsureState(request, WorkflowState.PendingManager);
        request.ApproverId = managerId;
        request.DecisionNote = note;
        request.DecidedAt = at;
        request.WorkflowState = WorkflowState.Rejected;
        request.Status = RequestStatus.Rejected;
        SyncDerived(request, at);
    }

    public static void HrConfirm(Request request, Guid hrAdminId, DateTimeOffset at, string? note = null)
    {
        EnsureState(request, WorkflowState.PendingHr);
        request.HrConfirmedAt = at;
        request.DecisionNote = note ?? request.DecisionNote;
        request.WorkflowState = WorkflowState.Approved;
        request.Status = RequestStatus.Approved;
        SyncDerived(request, at);
    }

    public static void HrReject(Request request, Guid hrAdminId, DateTimeOffset at, string? note = null)
    {
        EnsureState(request, WorkflowState.PendingHr);
        request.ApproverId = hrAdminId;
        request.DecisionNote = note;
        request.DecidedAt = at;
        request.WorkflowState = WorkflowState.Rejected;
        request.Status = RequestStatus.Rejected;
        SyncDerived(request, at);
    }

    public static void ReturnForRevision(Request request, Guid managerId, DateTimeOffset at, string note)
    {
        EnsureState(request, WorkflowState.PendingManager, WorkflowState.PendingSubstitute);
        if (string.IsNullOrWhiteSpace(note))
        {
            throw new InvalidOperationException("A note is required when returning a request for revision.");
        }
        request.WorkflowState = WorkflowState.ReturnedForRevision;
        request.Status = RequestStatus.Submitted;
        request.DecisionNote = note;
        SyncDerived(request, at);
    }

    public static void Cancel(Request request, Guid actorId, DateTimeOffset at, string? note = null)
    {
        if (request.WorkflowState is WorkflowState.Cancelled or WorkflowState.Rejected)
        {
            throw new InvalidOperationException(
                $"Request {request.Id} is in terminal state '{request.WorkflowState}' and cannot be cancelled.");
        }
        if (request.WorkflowState == WorkflowState.Approved && request.From <= at)
        {
            throw new InvalidOperationException(
                "Approved requests cannot be cancelled after the start date has passed.");
        }
        request.WorkflowState = WorkflowState.Cancelled;
        request.Status = RequestStatus.Rejected;
        request.CancelledAt = at;
        request.DecisionNote = note ?? request.DecisionNote;
        SyncDerived(request, at);
    }

    public static bool IsOpen(WorkflowState state) => state is
        WorkflowState.Draft or
        WorkflowState.Submitted or
        WorkflowState.PendingSubstitute or
        WorkflowState.PendingManager or
        WorkflowState.PendingHr or
        WorkflowState.ReturnedForRevision;

    private static void EnsureState(Request request, params WorkflowState[] allowed)
    {
        if (!allowed.Contains(request.WorkflowState))
        {
            throw new InvalidOperationException(
                $"Request {request.Id} is in workflow state '{request.WorkflowState}'. Allowed: {string.Join(", ", allowed)}.");
        }
    }

    private static void EnsureActor(Guid? expectedId, Guid actorId, string role)
    {
        if (expectedId is null || expectedId.Value != actorId)
        {
            throw new InvalidOperationException($"Actor {actorId} is not the {role} for this request.");
        }
    }

    private static void SyncDerived(Request request, DateTimeOffset at)
    {
        // CurrentApproverId is purely informational (UI routing); persistence layer fills it.
        if (request.WorkflowState is WorkflowState.Approved or WorkflowState.Rejected or WorkflowState.Cancelled)
        {
            request.CurrentApproverId = null;
        }
    }
}

File: libs/shared/src/lib/workflow/vacation-workflow.ts
export type WorkflowState =
  | 'Draft'
  | 'Submitted'
  | 'PendingSubstitute'
  | 'PendingManager'
  | 'PendingHr'
  | 'Approved'
  | 'Rejected'
  | 'Cancelled';

export type WorkflowEvent =
  | 'submit'
  | 'substitute_accept'
  | 'substitute_decline'
  | 'manager_approve'
  | 'manager_approve_with_hr'
  | 'manager_reject'
  | 'manager_return'
  | 'hr_confirm'
  | 'hr_reject'
  | 'cancel';

export class InvalidWorkflowTransitionError extends Error {
  readonly state: WorkflowState;
  readonly event: WorkflowEvent;
  constructor(state: WorkflowState, event: WorkflowEvent) {
    super(`Invalid workflow transition: cannot apply "${event}" in state "${state}"`);
    this.name = 'InvalidWorkflowTransitionError';
    this.state = state;
    this.event = event;
  }
}

export interface TransitionContext {
  hasSubstitute: boolean;
}

const FINAL_STATES: ReadonlySet<WorkflowState> = new Set(['Approved', 'Rejected', 'Cancelled']);

export function isFinal(state: WorkflowState): boolean {
  return FINAL_STATES.has(state);
}

export function nextState(
  state: WorkflowState,
  event: WorkflowEvent,
  ctx: TransitionContext = { hasSubstitute: false },
): WorkflowState {
  if (event === 'cancel') {
    if (isFinal(state)) throw new InvalidWorkflowTransitionError(state, event);
    return 'Cancelled';
  }

  switch (state) {
    case 'Draft':
    case 'Submitted':
      if (event === 'submit') {
        return ctx.hasSubstitute ? 'PendingSubstitute' : 'PendingManager';
      }
      break;
    case 'PendingSubstitute':
      if (event === 'substitute_accept') return 'PendingManager';
      if (event === 'substitute_decline') return 'Draft';
      break;
    case 'PendingManager':
      if (event === 'manager_approve') return 'Approved';
      if (event === 'manager_approve_with_hr') return 'PendingHr';
      if (event === 'manager_reject') return 'Rejected';
      if (event === 'manager_return') return 'Draft';
      break;
    case 'PendingHr':
      if (event === 'hr_confirm') return 'Approved';
      if (event === 'hr_reject') return 'Rejected';
      break;
    case 'Approved':
    case 'Rejected':
    case 'Cancelled':
      break;
  }
  throw new InvalidWorkflowTransitionError(state, event);
}

export type RequestStatus = 'Submitted' | 'Approved' | 'Rejected' | 'Cancelled';

export function deriveStatus(state: WorkflowState): RequestStatus {
  if (state === 'Approved') return 'Approved';
  if (state === 'Rejected') return 'Rejected';
  if (state === 'Cancelled') return 'Cancelled';
  return 'Submitted';
}
